app/src/main/java/com/alexafit/tasky/MainActivity.kt

```diff
@@ -3,49 +3,97 @@ package com.alexafit.tasky
+
+        lifecycleScope.launch {
+            lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
+                mainViewModel.uiState.collect {
+                    uiState = it
+                }
+            }
+        }
+        splashScreen.setKeepOnScreenCondition {
+            when (uiState) {
+                MainActivityUiState.Loading -> true
+                MainActivityUiState.Success -> false
+            }
+        }
         setContent {
             TaskyTheme {
+                val modifier = Modifier
+                Box(
+                    modifier = modifier.fillMaxSize(),
+                    contentAlignment = Alignment.Center
+                ) {}
                 val navController = rememberNavController()
                 val scaffoldState = rememberScaffoldState()
                 Scaffold(
-                    modifier = Modifier.fillMaxSize(),
+                    modifier = modifier.fillMaxSize(),
                     scaffoldState = scaffoldState
                 ) { padding ->
                     NavHost(
                         navController = navController,
-                        startDestination = if (activeSession) Route.AGENDA_OVERVIEW else Route.LOGIN,
-                        modifier = Modifier.padding(padding)
+                        startDestination = when (mainViewModel.activeSession.value) {
+                            true -> Route.AGENDA_OVERVIEW
+                            false -> Route.LOGIN
+                            else -> Route.LOADING_SCREEN
+                        },
+                        modifier = modifier.padding(padding)
                     ) {
+                        composable(Route.LOADING_SCREEN) {
+                            Box(
+                                modifier = modifier.fillMaxSize(),
+                                contentAlignment = Alignment.Center
+                            ) {
+                                CircularLoadingDialog(modifier = modifier)
+                            }
+                        }
                         composable(Route.LOGIN) {
                             LoginScreen(
                                 scaffoldState = scaffoldState,
                                 onEvent = { event ->
                                     when (event) {
-                                        NavigationEvent.NavigateToAgenda -> navController.navigate(Route.AGENDA_OVERVIEW)
-                                        NavigationEvent.NavigateToRegister -> navController.navigate(Route.REGISTER)
+                                        NavigationEvent.NavigateToAgenda -> navController.navigate(
+                                            Route.AGENDA_OVERVIEW
+                                        )
+                                        NavigationEvent.NavigateToRegister -> navController.navigate(
+                                            Route.REGISTER
+                                        )
```

Comment:
I personally like to have this handled on a LoginViewModel

For register, it would be on a RegisterViewModel